### DISCUSSION TITLE
Any way to speed up matrix multiplication (and similar queries with different aggregates)?

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#11563] [General]
Data transformations that can be rephrased as matrix multiplications are currently many orders of magnitude faster in BLAS/numpy/pandas than in duckdb. For example, mulitplying two matrices of size `2000x2000` takes 25s in duckdb but only 27ms using pandas' `@` operator -- a 900x speedup (for `4000x4000` the difference becomes 220s / 205ms = 1100x; see benchmark code below).

Obviously it's expected that BLAS is faster at what it's specifically designed for, but often moving to numpy for intermediate steps can be a big hassle (and requires hacks for missing data / different shapes per different group keys / ....) so I'd be greatly interested if there is something I could do (or whether duckdb could be improved with this use pattern in mind) to speed things up to say a 10x-50x slowdown compared to BLAS, which I got with a very vanilla C# hash-based sparse matrix multiplication implementation so should be achievable with duckdb as well.

By "can be phrased as matrix multiplication" I mean SQL code that looks like 

```sql
SELECT 
a.key,
b.key,
sum(a.value * b.value)
FROM a
JOIN b
ON a.match = b.match
GROUP BY a.key, b.key
```

Presumably any speed gains for this pattern will also make the same query with different aggregates faster.

--- 

Benchmark:

```python
import duckdb
import pandas as pd
import numpy as np

n = 2000
ar = pd.DataFrame(np.random.rand(n * n).reshape([n, n]))
a = ar.stack().reset_index()
br = pd.DataFrame(np.random.rand(n * n).reshape([n, n]))
b = br.stack().reset_index()

# %%timeit
m_pandas = ar @ br.T

# %%timeit
m_duckdb = duckdb.query("""
SELECT 
a.level_0 AS key1,
b.level_0 AS key2,
sum(a."0" * b."0")
FROM a
JOIN b
ON a.level_1 = b.level_1
GROUP BY key1, key2
""").df().set_index(['key1', 'key2']).unstack('key2')

assert np.abs(m_duckdb.values - m_pandas.values).sum() < 1e-5
```

I didn't spend time optimizing away things like the pandas conversion and pre and postprocessing or the ordering, since duckdb spends 90% of the time in 
```
       HASH_GROUP_BY       │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│             #0            │                             
│             #1            │                             
│          sum(#2)          │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│          1000000          │                             
│         (151.52s)         
```

In comparison, a completely unoptimized "hashed sparse" matrix multiply of two dense 2000x2000 matrices in C# using matrices stored as `(int, int, double)[]`, using `Dictionary<int, (int, double)[]>` for hashing of `a` and `Dictionary<int, Dictionary<int, double>` for hashing of `b`, `.AsParallel().Select()` to parallelize over the groups, and a manual group-wise accumulation using a for loop with `secondMatrixValuesInGroup.TryGet(firstMatrixKey, out var x)` takes 3.8seconds only (of which 0.5 seconds is creating the matrices and datastructures), a 7x speedup over duckdb.
With some fine-tuning (which still works for non-aligned matrices but might not be faster for them because it aligns the values of the columns that are being matched in a join across all groups), this can be reduced to 1.2s (of which still 0.5 seconds are creating matrices and datastructures), a 21x speedup over duckdb and less then 50x slower than BLAS.

PS: I'm multiplying `a @ b.T` instead of `a @ b` because of https://github.com/duckdb/duckdb/issues/11568 ; otherwise, the difference is 3000x and duckdb can only do very small matrices without running out of memory.

**Comment by szarnyasg** [comment]
Hi @soerenwolfers, thanks for opening this thread. One idea that could work is using [`POSITIONAL JOIN`](https://duckdb.org/docs/sql/query_syntax/from#positional-joins).

A personal remark: matrix multiplication is a topic close to my heart as I worked quite a lot on algorithms related to GraphBLAS (sparse matrices for graph processing). So I'm excited to see what DuckDB can do with vanilla SQL for matrix multiplication.

**Reply by soerenwolfers** [reply]
Could you give an example how a positional join could be used? The JOIN i was using produces $n^3$ rows, but positional joins can only produce as many rows as the inputs, no?

Glad you're excited for this too. Simplicity of sparse matrix multiplication, especially the storage which comes for free in ducks, is ultimately what duckdb will shine at, but if I am to use duckdb for this, I feel like starting out with a 1000x disadvantage at dense matrices is a quite big hurdle to overcome.

**Comment by soerenwolfers** [comment]
I created an issue at https://github.com/duckdb/duckdb/issues/11588 with some more evidence that something might be wrong

**Reply by soerenwolfers** [reply]
On that ticket it was concluded that there is nothing obvious to be done in the short term since the query optimizer doesn't typically optimize across join-and-aggregate and exploding joins are not common.

**Reply by FrancoisLepoutre** [reply]
Thanks for feeding back

**Comment by FrancoisLepoutre** [comment]
Hi

_matrix multiplications are currently many orders of magnitude faster in BLAS/numpy/pandas than in duckdb._

Thanks for this useful thread. I do feel alone trying to handle matrix-based operations within the duckdb engine :-)

For sure, Duckdb cannot provide the impressive speed that BLAS/numpy/pandas and others may provide. May i add to the list of numpy-based data crunching alternatives, numba and aot solutions such as pythran and others.

In the end, as far as I am concerned, I have discarded the idea that our beloved duck is the best fit for massive in-memory matrix-based calculation sets. The last calculation mile is intended for the python-numpy combo. When one understands what is at stake in terms cpu-handling in a relational db engine, that's of course plenty obvious. But we always try to feed the duckb whenever possible and that's the good starting point until speed becomes an issue. And it becomes one when you drift into matrix operations. After P-O-C-ing a few alternative, our coming app will:
-  feed the data from our beloved duck to python as a numpy structure,
- run the data-crunching within the numpy ecosystem,
- feed the data back to our duck.

I understand that most exchanges between numpy or pyarrow structures and duckdb tend to be zero-copy. So may we say we possibly use these tools in what they do best:

- duckdb, "the storage", in charge of both persistent and in-memory storage including sorting, subsetting and binning of massive datasets,
- the numpy eco-system, "the hammer" : in charge of massive on-the-fly calculations when required.

**Reply by soerenwolfers** [reply]
What's "our coming app"? An ibis style wrapper around duckdb? A static function of the form `einsum(duckdbtable1, duckdbtable2, dimensions)`?

**Reply by FrancoisLepoutre** [reply]
_What's "our coming app"?_

A highly specialized desktop app targeted at non-tech end-users. Not yet started and still specifying. Takes time... But sure, the duckdb will be one of its core "enabling" components along with a couple of python-based libs. Thanks to all contributors to the our great duck!

_A static function of the form einsum(duckdbtable1, duckdbtable2, dimensions)?_

By trade I am a db guy. Not really on the matrix side of IT life :-) But I'd be glad to hear anyone considering some sort of embedded numpy-like set of matrix handling functions straight within our duck. Should that be possible!

I have little understanding of the internals of duckdb. And that's possibly why I doubt you could reach the kind of perf that simpler and highly compact data structures à la numpy can provide. But I am a dummy on those issues.
